### ISSUE TITLE
Overriding classes with Injector breaks MTI

### ISSUE THREAD

**Issue by dhensby** [silverstripe/silverstripe-framework#9950]
## Affected Version

4

## Description

see https://github.com/silverstripe/silverstripe-userforms/pull/1060 for some context

When overriding a DataObject class using a child class and Injector to replace it, the multi-table inheritance is broken.

The underlying DB is not built and the table inheritance does not get performed.

This comes about because the logic to build the database schema uses `singleton($class)`, which uses injector to then load the child class and not the original class. Then `$sng->requireTable()` uses late static binding to build the table for the overridden class instead of for the original one.

I suspect this is a regression from when `DataObjectSchema` was introduced.

Looking at this code: https://github.com/silverstripe/silverstripe-framework/blob/4.7.3/src/ORM/DatabaseAdmin.php#L304-L344

---

```php
        $dbSchema = DB::get_schema();
        $dbSchema->schemaUpdate(function () use ($dataClasses, $testMode, $quiet, $showRecordCounts) {
            $dataObjectSchema = DataObject::getSchema();

            foreach ($dataClasses as $dataClass) {
                // Check if class exists before trying to instantiate - this sidesteps any manifest weirdness
                if (!class_exists($dataClass)) {
                    continue;
                }

                // Check if this class should be excluded as per testing conventions
                $SNG = singleton($dataClass);
                if (!$testMode && $SNG instanceof TestOnly) {
                    continue;
                }
                // here tableName is fetched for the original `$dataClass`, which is good.
                $tableName = $dataObjectSchema->tableName($dataClass);

                // Log data
                if (!$quiet) {
                    if ($showRecordCounts && DB::get_schema()->hasTable($tableName)) {
                        try {
                            $count = DB::query("SELECT COUNT(*) FROM \"$tableName\"")->value();
                            $countSuffix = " ($count records)";
                        } catch (Exception $e) {
                            $countSuffix = " (error getting record count)";
                        }
                    } else {
                        $countSuffix = "";
                    }

                    if (Director::is_cli()) {
                        echo " * $tableName$countSuffix\n";
                    } else {
                        echo "<li>$tableName$countSuffix</li>\n";
                    }
                }

                // Instruct the class to apply its schema to the database
                // here the `requireTable` call happens on the injector created instance instead of the original class
                $SNG->requireTable();
            }
        });
```

---

## Steps to Reproduce

Install userdefinedforms and set up a class like so:

```yml
SilverStripe\Core\Injector\Injector:
  SilverStripe\UserForms\Model\Submission\SubmittedFileField:
    class: MySubmittedFileField
```

```php
<?php

namespace {

    use SilverStripe\ORM\FieldType\DBField;
    use SilverStripe\UserForms\Model\Submission\SubmittedFileField;

    class MySubmittedFileField extends SubmittedFileField {

        /**
         * Return the value of this field for inclusion into things such as
         * reports.
         *
         * @return string
         */
        public function getFormattedValue()
        {
            $name = $this->getFileName();
            $link = $this->getLink();
            $title = 'Extended Download File';

            if ($link) {
                return DBField::create_field('HTMLText', sprintf(
                    '%s - <a href="%s" target="_blank">%s</a>',
                    $name,
                    $link,
                    $title
                ));
            }

            return false;
        }
    }
}
```

# Acceptance criteria
- [ ] Make a reasonable effort to write a unit test for a DataObject that extends and override its parent.

# Notes
We have a PR that seems to fix this, but we are not sure this can be tested.

Unit test idea: skip if current database is not mysql. check table exists in raw query - SiteTree + MySiteTree + relevant columns 

# Pull request
- [ ] https://github.com/silverstripe/silverstripe-framework/pull/9952

**Comment by michalkleiner** [comment]
Another dimension to this is that the overriding class needs to use the same $db table, if it doesn't specify its own one via `$table_name`. If it does specify its own one, it currently can't be the same one as the one from the original class (Config complains about that, I guess for a good reason).

**Comment by dhensby** [comment]
@michalkleiner could you try out my very basic fix in #9951 and see if that works and if it has any unintended side-affects?

**Comment by maxime-rainville** [comment]
Looks like we might be able to fix it with this very simple PR https://github.com/silverstripe/silverstripe-framework/pull/9952
